statesync/dispatcher_test.go

```diff
@@ -0,0 +1,160 @@
+package statesync
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tendermint/tendermint/p2p"
+	ssproto "github.com/tendermint/tendermint/proto/tendermint/statesync"
+)
+
+func TestDispatcherBasic(t *testing.T) {
+
+	ch := make(chan p2p.Envelope, 100)
+	closeCh := make(chan struct{})
+	defer close(closeCh)
+
+	d := newDispatcher(ch, 1*time.Second)
+
+	go handleRequests(t, d, ch, closeCh)
+
+	peers := createPeerSet(5)
+	for _, peer := range peers {
+		d.addPeer(peer)
+	}
+
+	wg := sync.WaitGroup{}
+
+	// make a bunch of async requests and require that the correct responses are
+	// given
+	for i := 1; i < 10; i++ {
+		wg.Add(1)
+		go func(height int64) {
+			lb, peer, err := d.LightBlock(context.Background(), height)
+			require.NoError(t, err)
+			require.NotNil(t, lb)
+			require.Equal(t, lb.Height, height)
+			require.Contains(t, peers, peer)
+			wg.Done()
+		}(int64(i))
+	}
+	wg.Wait()
+}
+
+func TestDispatcherProviders(t *testing.T) {
+
+	ch := make(chan p2p.Envelope, 100)
+	chainID := "state-sync-test"
+	closeCh := make(chan struct{})
+	defer close(closeCh)
+
+	d := newDispatcher(ch, 1*time.Second)
+
+	go handleRequests(t, d, ch, closeCh)
+
+	peers := createPeerSet(5)
+	for _, peer := range peers {
+		d.addPeer(peer)
+	}
+
+	providers := d.Providers(chainID, 5*time.Second)
+	assert.Len(t, providers, 5)
+	for i, p := range providers {
+		bp, ok := p.(*blockProvider)
+		require.True(t, ok)
+		assert.Equal(t, bp.String(), string(peers[i]))
+		lb, err := p.LightBlock(context.Background(), 10)
+		assert.Error(t, err)
+		assert.Nil(t, lb)
+	}
+}
+
+func TestPeerListBasic(t *testing.T) {
+	peerList := newPeerList()
+	assert.Zero(t, peerList.Len())
+	numPeers := 10
+	peerSet := createPeerSet(numPeers)
+
+	for _, peer := range peerSet {
+		peerList.Append(peer)
+	}
+
+	for idx, peer := range peerList.Peers() {
+		assert.Equal(t, peer, peerSet[idx])
+	}
+
+	assert.Equal(t, numPeers, peerList.Len())
+
+	half := numPeers / 2
+	for i := 0; i < half; i++ {
+		assert.Equal(t, peerSet[i], peerList.Pop())
+	}
+	assert.Equal(t, half, peerList.Len())
+
+	peerList.Remove(p2p.NodeID("lp"))
+	assert.Equal(t, half, peerList.Len())
+
+	peerList.Remove(peerSet[half])
+	half++
+	assert.Equal(t, peerSet[half], peerList.Pop())
+
+}
+
+func TestPeerListConcurrent(t *testing.T) {
```

Comment:
is this just "assert that nothing panics?" should there be an assertion?

Reply:
I'm trying to check that the go routines wait until a peer is added to the set and that it won't lock up

Reply:
is there a way to do that isn't "the test will panic" or "the test will deadlock/timeout?" Is there a way we can make that more explicit in the test code just so when we look at it in the future it's clear?

Reply:
Sure I can add a timeout and some comments